crates/host/src/policy.rs

```diff
@@ -0,0 +1,354 @@
+use core::time::Duration;
+
+use std::collections::{hash_map, HashMap};
+use std::hash::Hash;
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use anyhow::Context;
+use futures::{
+    stream::{AbortHandle, Abortable},
+    StreamExt,
+};
+use serde::{Deserialize, Serialize, Serializer};
+use tokio::spawn;
+use tokio::sync::RwLock;
+use tracing::{debug, error, instrument, trace, warn};
+use ulid::Ulid;
+use uuid::Uuid;
+use wascap::jwt;
+
+/// Relevant information about the actor or provider making an invocation. This struct is empty for
+/// policy decisions related to starting actors or providers. All fields are optional for backwards-compatibility
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Hash)]
+pub struct RequestSource {
+    /// The public key of the actor or provider
+    #[serde(rename = "publicKey")]
+    pub public_key: Option<String>,
+    /// The contract ID of the provider, or None if the source is an actor
+    #[serde(rename = "contractId")]
+    pub contract_id: Option<String>,
+    /// The link name of the provider, or None if the source is an actor
+    #[serde(rename = "linkName")]
+    pub link_name: Option<String>,
+    /// The list of capabilities of the actor, or None if the source is a provider
+    pub capabilities: Vec<String>,
+    /// The issuer of the source's claims
+    pub issuer: Option<String>,
+    /// The time the claims were signed
+    #[serde(rename = "issuedOn")]
+    pub issued_on: Option<String>,
+    /// The time the claims expire, if any
+    #[serde(rename = "expiresAt")]
+    pub expires_at: Option<u64>,
+    /// Whether the claims have expired already. This is included in case the policy server is fulfilled by an actor, which cannot access the system clock
+    pub expired: bool,
+}
+
+/// Relevant information about the actor that is being invoked, or the actor or provider that is
+/// being started. All fields are optional for backwards-compatibility
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Hash)]
+pub struct RequestTarget {
```

Comment:
Why is this not an enum? (I'm fine with a TODO to change it if we need custom `Serialize` logic for that)

Reply:
It should be, but we do have to serialize all the fields (as you noted) for backwards-compatibility with existing implementations of a policy service. I've added TODOs to `RequestTarget` and `RequestSource` to reflect that we want to change this 👍 

Original RFC for reference https://github.com/wasmCloud/wasmcloud-otp/issues/439